# Assignment 3

By Valentin PICAVET

# Question 1
Prepare the data for this assignment: 
- Make sure you have already imported “pandas” and “yfinance” into Python.
- Download from Wikipedia (or any other source) a table that lists the companies that comprise the S&P 500. (See “Helpful links” provided at the end of the assignment.)
- From this table, extract the list of ticker symbols (short names for all the companies).
- Set the start date and end date to be
- start date = ”2000-01-01”
- end date = ”2022-12-31”
- Build a dataframe that contains the stock prices for the S&P 500 companies. (If there are errors for some company names, it is fine to ignore the company names with errors.)
- Drop the columns that have only “NaN” entries.
- Drop also the company names that have more than 100 missing observations.

In [1]:
import yfinance as yf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
URL = 'https://en.wikipedia.org/wiki/List_of_S%26P_500_companies' # Wikipedia Webpage

L_tickers = pd.read_html(URL)[0]['Symbol'].tolist() #List Containing the Ticker names

# Define start and end dates
start_date = '2000-01-01'
end_date = '2022-12-31'

# Fetch adjusted close prices for the selected stocks
stock_data = yf.download(L_tickers, start=start_date, end=end_date)['Adj Close']


stock_data

[*********************100%%**********************]  503 of 503 completed


4 Failed downloads:
['VLTO', 'KVUE']: Exception("%ticker%: Data doesn't exist for startDate = 946702800, endDate = 1672462800")
['BF.B']: Exception('%ticker%: No price data found, symbol may be delisted (1d 2000-01-01 -> 2022-12-31)')
['BRK.B']: Exception('%ticker%: No timezone found, symbol may be delisted')


A    AAL        AAPL        ABBV       ABNB         ABT  \
Date                                                                           
2000-01-03   43.683865    NaN    0.846127         NaN        NaN    9.037448   
2000-01-04   40.346901    NaN    0.774790         NaN        NaN    8.779236   
2000-01-05   37.844177    NaN    0.786128         NaN        NaN    8.763097   
2000-01-06   36.403214    NaN    0.718098         NaN        NaN    9.069723   
2000-01-07   39.436829    NaN    0.752113         NaN        NaN    9.166555   
...                ...    ...         ...         ...        ...         ...   
2022-12-23  147.932922  12.71  130.959961  155.283295  85.250000  105.548424   
2022-12-27  148.250122  12.53  129.142441  155.178574  83.489998  105.928925   
2022-12-28  146.802811  12.32  125.179680  154.455002  82.489998  105.206917   
2022-12-29  149.776733  12.70  128.725311  154.769180  85.230003  107.626587   
2022-12-30  148.570496  12.72  129.043106  153.864700  85.500000  107.119255   

                 ACGL         ACN        ADBE         ADI  ...         WY  \
Date                                                       ...              
2000-01-03   1.277778         NaN   16.274673   28.438295  ...  11.505334   
2000-01-04   1.270833         NaN   14.909399   26.999620  ...  11.073109   
2000-01-05   1.388889         NaN   15.204175   27.393782  ...  11.659705   
2000-01-06   1.375000         NaN   15.328291   26.644886  ...  12.205119   
2000-01-07   1.451389         NaN   16.072983   27.393782  ...  11.803772   
...               ...         ...         ...         ...  ...        ...   
2022-12-23  63.380001  260.975525  338.450012  160.024155  ...  29.621927   
2022-12-27  63.619999  260.210541  335.089996  158.422363  ...  29.603115   
2022-12-28  62.599998  258.062622  328.329987  156.547058  ...  28.812946   
2022-12-29  63.110001  263.221558  337.579987  160.160904  ...  29.462011   
2022-12-30  62.779999  261.711121  336.529999  160.209732  ...  29.160995   

                 WYNN        XEL         XOM       XRAY         XYL  \
Date                                                                  
2000-01-03        NaN   6.977993   18.328690   6.669164         NaN   
2000-01-04        NaN   7.138673   17.977627   6.669164         NaN   
2000-01-05        NaN   7.414119   18.957701   6.784459         NaN   
2000-01-06        NaN   7.345262   19.937757   6.793326         NaN   
2000-01-07        NaN   7.345262   19.879250   6.775589         NaN   
...               ...        ...         ...        ...         ...   
2022-12-23  79.901894  67.391136  104.124237  31.219860  108.006180   
2022-12-27  83.475304  67.999199  105.570953  31.455261  108.980629   
2022-12-28  79.278267  67.509399  103.836830  30.386158  107.228592   
2022-12-29  80.436424  67.987724  104.622444  31.789503  109.886169   
2022-12-30  81.634148  67.069351  105.676338  31.356190  108.832985   

                   YUM         ZBH        ZBRA         ZTS  
Date                                                        
2000-01-03    4.680298         NaN   25.027779         NaN  
2000-01-04    4.586222         NaN   24.666668         NaN  
2000-01-05    4.609739         NaN   25.138889         NaN  
2000-01-06    4.570545         NaN   23.777779         NaN  
2000-01-07    4.468628         NaN   23.513889         NaN  
...                ...         ...         ...         ...  
2022-12-23  125.889427  125.714478  248.220001  144.115540  
2022-12-27  126.866081  126.299927  251.000000  143.660751  
2022-12-28  126.289864  125.019867  246.839996  142.207306  
2022-12-29  126.953979  126.845688  257.529999  146.478577  
2022-12-30  125.088593  126.518242  256.410004  144.896622  

[5787 rows x 503 columns]

- Drop the columns that have only “NaN” entries.

In [3]:
stocks_prices=stock_data

In [4]:
stocks_prices.dropna(axis=1, how='all', inplace=True)

In [5]:
stocks_prices

A    AAL        AAPL        ABBV       ABNB         ABT  \
Date                                                                           
2000-01-03   43.683865    NaN    0.846127         NaN        NaN    9.037448   
2000-01-04   40.346901    NaN    0.774790         NaN        NaN    8.779236   
2000-01-05   37.844177    NaN    0.786128         NaN        NaN    8.763097   
2000-01-06   36.403214    NaN    0.718098         NaN        NaN    9.069723   
2000-01-07   39.436829    NaN    0.752113         NaN        NaN    9.166555   
...                ...    ...         ...         ...        ...         ...   
2022-12-23  147.932922  12.71  130.959961  155.283295  85.250000  105.548424   
2022-12-27  148.250122  12.53  129.142441  155.178574  83.489998  105.928925   
2022-12-28  146.802811  12.32  125.179680  154.455002  82.489998  105.206917   
2022-12-29  149.776733  12.70  128.725311  154.769180  85.230003  107.626587   
2022-12-30  148.570496  12.72  129.043106  153.864700  85.500000  107.119255   

                 ACGL         ACN        ADBE         ADI  ...         WY  \
Date                                                       ...              
2000-01-03   1.277778         NaN   16.274673   28.438295  ...  11.505334   
2000-01-04   1.270833         NaN   14.909399   26.999620  ...  11.073109   
2000-01-05   1.388889         NaN   15.204175   27.393782  ...  11.659705   
2000-01-06   1.375000         NaN   15.328291   26.644886  ...  12.205119   
2000-01-07   1.451389         NaN   16.072983   27.393782  ...  11.803772   
...               ...         ...         ...         ...  ...        ...   
2022-12-23  63.380001  260.975525  338.450012  160.024155  ...  29.621927   
2022-12-27  63.619999  260.210541  335.089996  158.422363  ...  29.603115   
2022-12-28  62.599998  258.062622  328.329987  156.547058  ...  28.812946   
2022-12-29  63.110001  263.221558  337.579987  160.160904  ...  29.462011   
2022-12-30  62.779999  261.711121  336.529999  160.209732  ...  29.160995   

                 WYNN        XEL         XOM       XRAY         XYL  \
Date                                                                  
2000-01-03        NaN   6.977993   18.328690   6.669164         NaN   
2000-01-04        NaN   7.138673   17.977627   6.669164         NaN   
2000-01-05        NaN   7.414119   18.957701   6.784459         NaN   
2000-01-06        NaN   7.345262   19.937757   6.793326         NaN   
2000-01-07        NaN   7.345262   19.879250   6.775589         NaN   
...               ...        ...         ...        ...         ...   
2022-12-23  79.901894  67.391136  104.124237  31.219860  108.006180   
2022-12-27  83.475304  67.999199  105.570953  31.455261  108.980629   
2022-12-28  79.278267  67.509399  103.836830  30.386158  107.228592   
2022-12-29  80.436424  67.987724  104.622444  31.789503  109.886169   
2022-12-30  81.634148  67.069351  105.676338  31.356190  108.832985   

                   YUM         ZBH        ZBRA         ZTS  
Date                                                        
2000-01-03    4.680298         NaN   25.027779         NaN  
2000-01-04    4.586222         NaN   24.666668         NaN  
2000-01-05    4.609739         NaN   25.138889         NaN  
2000-01-06    4.570545         NaN   23.777779         NaN  
2000-01-07    4.468628         NaN   23.513889         NaN  
...                ...         ...         ...         ...  
2022-12-23  125.889427  125.714478  248.220001  144.115540  
2022-12-27  126.866081  126.299927  251.000000  143.660751  
2022-12-28  126.289864  125.019867  246.839996  142.207306  
2022-12-29  126.953979  126.845688  257.529999  146.478577  
2022-12-30  125.088593  126.518242  256.410004  144.896622  

[5787 rows x 499 columns]

-Drop also the company names that have more than 100 missing observations.

In [6]:
for stock in stocks_prices.columns:
    s=stocks_prices[stock].isna().sum() # Count the number of NaN
    if s>100:
        stocks_prices=stocks_prices.drop(columns=stock)
stocks_prices

A        AAPL         ABT       ACGL        ADBE  \
Date                                                                    
2000-01-03   43.683865    0.846127    9.037448   1.277778   16.274673   
2000-01-04   40.346901    0.774790    8.779236   1.270833   14.909399   
2000-01-05   37.844177    0.786128    8.763097   1.388889   15.204175   
2000-01-06   36.403214    0.718098    9.069723   1.375000   15.328291   
2000-01-07   39.436829    0.752113    9.166555   1.451389   16.072983   
...                ...         ...         ...        ...         ...   
2022-12-23  147.932922  130.959961  105.548424  63.380001  338.450012   
2022-12-27  148.250122  129.142441  105.928925  63.619999  335.089996   
2022-12-28  146.802811  125.179680  105.206917  62.599998  328.329987   
2022-12-29  149.776733  128.725311  107.626587  63.110001  337.579987   
2022-12-30  148.570496  129.043106  107.119255  62.779999  336.529999   

                   ADI        ADM         ADP        ADSK        AEE  ...  \
Date                                                                  ...   
2000-01-03   28.438295   6.413243   24.781528    8.052905  10.672665  ...   
2000-01-04   26.999620   6.346437   24.781528    7.660816  10.672665  ...   
2000-01-05   27.393782   6.246231   24.543249    7.178246  11.085534  ...   
2000-01-06   26.644886   6.279636   24.870886    6.740915  11.044241  ...   
2000-01-07   27.393782   6.379841   25.436806    7.540173  11.209390  ...   
...                ...        ...         ...         ...        ...  ...   
2022-12-23  160.024155  91.612831  234.232559  188.160004  85.987610  ...   
2022-12-27  158.422363  92.861839  234.320053  186.289993  86.707939  ...   
2022-12-28  156.547058  90.644615  231.228577  181.899994  85.843544  ...   
2022-12-29  160.160904  90.160515  233.921448  188.119995  86.448624  ...   
2022-12-30  160.209732  89.899078  232.210449  186.869995  85.401749  ...   

                  WMB        WMT        WRB         WST         WY        XEL  \
Date                                                                            
2000-01-03   9.374685  14.469357   1.226067    5.428276  11.505334   6.977993   
2000-01-04   9.166777  13.927946   1.188342    5.484353  11.073109   7.138673   
2000-01-05   9.733794  13.643707   1.169480    5.461923  11.659705   7.414119   
2000-01-06   9.941700  13.792596   1.237385    5.405844  12.205119   7.345262   
2000-01-07  10.206306  14.834825   1.218522    5.461923  11.803772   7.345262   
...               ...        ...        ...         ...        ...        ...   
2022-12-23  31.123585  47.043129  71.307037  234.454605  29.621927  67.391136   
2022-12-27  31.207653  47.056225  71.753624  236.050003  29.603115  67.999199   
2022-12-28  30.563137  46.231651  70.676003  230.575775  28.812946  67.509399   
2022-12-29  30.796654  46.513050  71.103157  238.293549  29.462011  67.987724   
2022-12-30  30.731272  46.395256  70.452713  234.673981  29.160995  67.069351   

                   XOM       XRAY         YUM        ZBRA  
Date                                                       
2000-01-03   18.328690   6.669164    4.680298   25.027779  
2000-01-04   17.977627   6.669164    4.586222   24.666668  
2000-01-05   18.957701   6.784459    4.609739   25.138889  
2000-01-06   19.937757   6.793326    4.570545   23.777779  
2000-01-07   19.879250   6.775589    4.468628   23.513889  
...                ...        ...         ...         ...  
2022-12-23  104.124237  31.219860  125.889427  248.220001  
2022-12-27  105.570953  31.455261  126.866081  251.000000  
2022-12-28  103.836830  30.386158  126.289864  246.839996  
2022-12-29  104.622444  31.789503  126.953979  257.529999  
2022-12-30  105.676338  31.356190  125.088593  256.410004  

[5787 rows x 362 columns]

# Question 2

- Compute the log returns for the companies in your dataset.

In [7]:
log_return_df=np.log(stocks_prices / stocks_prices.shift())

In [8]:
log_return_df

A      AAPL       ABT      ACGL      ADBE       ADI  \
Date                                                                     
2000-01-03       NaN       NaN       NaN       NaN       NaN       NaN   
2000-01-04 -0.079464 -0.088077 -0.028987 -0.005450 -0.087618 -0.051914   
2000-01-05 -0.064037  0.014527 -0.001840  0.088832  0.019578  0.014493   
2000-01-06 -0.038820 -0.090514  0.034392 -0.010050  0.008130 -0.027719   
2000-01-07  0.080043  0.046281  0.010620  0.054067  0.047440  0.027719   
...              ...       ...       ...       ...       ...       ...   
2022-12-23  0.001475 -0.002802  0.001388  0.008397  0.005719  0.000549   
2022-12-27  0.002142 -0.013976  0.003599  0.003779 -0.009977 -0.010060   
2022-12-28 -0.009811 -0.031166 -0.006839 -0.016163 -0.020380 -0.011908   
2022-12-29  0.020055  0.027931  0.022739  0.008114  0.027783  0.022822   
2022-12-30 -0.008086  0.002466 -0.004725 -0.005243 -0.003115  0.000305   

                 ADM       ADP      ADSK       AEE  ...       WMB       WMT  \
Date                                                ...                       
2000-01-03       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2000-01-04 -0.010471  0.000000 -0.049914  0.000000  ... -0.022427 -0.038136   
2000-01-05 -0.015915 -0.009662 -0.065063  0.037955  ...  0.060018 -0.020619   
2000-01-06  0.005334  0.013261 -0.062859 -0.003732  ...  0.021134  0.010854   
2000-01-07  0.015831  0.022499  0.112049  0.014843  ...  0.026268  0.072846   
...              ...       ...       ...       ...  ...       ...       ...   
2022-12-23  0.012014  0.007415  0.000319  0.013154  ...  0.022766  0.002019   
2022-12-27  0.013541  0.000373 -0.009988  0.008342  ...  0.002697  0.000278   
2022-12-28 -0.024166 -0.013281 -0.023848 -0.010019  ... -0.020869 -0.017679   
2022-12-29 -0.005355  0.011579  0.033623  0.007024  ...  0.007611  0.006068   
2022-12-30 -0.002904 -0.007341 -0.006667 -0.012184  ... -0.002125 -0.002536   

                 WRB       WST        WY       XEL       XOM      XRAY  \
Date                                                                     
2000-01-03       NaN       NaN       NaN       NaN       NaN       NaN   
2000-01-04 -0.031252  0.010278 -0.038291  0.022766 -0.019340  0.000000   
2000-01-05 -0.016000 -0.004098  0.051619  0.037859  0.053082  0.017140   
2000-01-06  0.056441 -0.010320  0.045717 -0.009331  0.050405  0.001306   
2000-01-07 -0.015361  0.010320 -0.033436  0.000000 -0.002939 -0.002614   
...              ...       ...       ...       ...       ...       ...   
2022-12-23  0.006419 -0.000553  0.014071  0.012770  0.026101  0.011057   
2022-12-27  0.006243  0.006782 -0.000635  0.008982  0.013798  0.007512   
2022-12-28 -0.015132 -0.023464 -0.027055 -0.007229 -0.016563 -0.034579   
2022-12-29  0.006026  0.032924  0.022277  0.007060  0.007537  0.045149   
2022-12-30 -0.009190 -0.015306 -0.010270 -0.013600  0.010023 -0.013724   

                 YUM      ZBRA  
Date                            
2000-01-03       NaN       NaN  
2000-01-04 -0.020305 -0.014533  
2000-01-05  0.005115  0.018963  
2000-01-06 -0.008539 -0.055665  
2000-01-07 -0.022551 -0.011160  
...              ...       ...  
2022-12-23  0.000621  0.002864  
2022-12-27  0.007728  0.011137  
2022-12-28 -0.004552 -0.016713  
2022-12-29  0.005245  0.042396  
2022-12-30 -0.014802 -0.004358  

[5787 rows x 362 columns]

# Question 3

- Compute the annual mean return, volatility, and Sharpe ratios for these companies in your dataset.

In [9]:
# Here we add a columns Year in our dataframe
L_year =[]
for date in log_return_df.index:
    L_year.append(date.year)

log_return_df['Year']=L_year
log_return_df

C:\Users\valou\AppData\Local\Temp\ipykernel_13096\4087453164.py:6: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  log_return_df['Year']=L_year


A      AAPL       ABT      ACGL      ADBE       ADI  \
Date                                                                     
2000-01-03       NaN       NaN       NaN       NaN       NaN       NaN   
2000-01-04 -0.079464 -0.088077 -0.028987 -0.005450 -0.087618 -0.051914   
2000-01-05 -0.064037  0.014527 -0.001840  0.088832  0.019578  0.014493   
2000-01-06 -0.038820 -0.090514  0.034392 -0.010050  0.008130 -0.027719   
2000-01-07  0.080043  0.046281  0.010620  0.054067  0.047440  0.027719   
...              ...       ...       ...       ...       ...       ...   
2022-12-23  0.001475 -0.002802  0.001388  0.008397  0.005719  0.000549   
2022-12-27  0.002142 -0.013976  0.003599  0.003779 -0.009977 -0.010060   
2022-12-28 -0.009811 -0.031166 -0.006839 -0.016163 -0.020380 -0.011908   
2022-12-29  0.020055  0.027931  0.022739  0.008114  0.027783  0.022822   
2022-12-30 -0.008086  0.002466 -0.004725 -0.005243 -0.003115  0.000305   

                 ADM       ADP      ADSK       AEE  ...       WMT       WRB  \
Date                                                ...                       
2000-01-03       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2000-01-04 -0.010471  0.000000 -0.049914  0.000000  ... -0.038136 -0.031252   
2000-01-05 -0.015915 -0.009662 -0.065063  0.037955  ... -0.020619 -0.016000   
2000-01-06  0.005334  0.013261 -0.062859 -0.003732  ...  0.010854  0.056441   
2000-01-07  0.015831  0.022499  0.112049  0.014843  ...  0.072846 -0.015361   
...              ...       ...       ...       ...  ...       ...       ...   
2022-12-23  0.012014  0.007415  0.000319  0.013154  ...  0.002019  0.006419   
2022-12-27  0.013541  0.000373 -0.009988  0.008342  ...  0.000278  0.006243   
2022-12-28 -0.024166 -0.013281 -0.023848 -0.010019  ... -0.017679 -0.015132   
2022-12-29 -0.005355  0.011579  0.033623  0.007024  ...  0.006068  0.006026   
2022-12-30 -0.002904 -0.007341 -0.006667 -0.012184  ... -0.002536 -0.009190   

                 WST        WY       XEL       XOM      XRAY       YUM  \
Date                                                                     
2000-01-03       NaN       NaN       NaN       NaN       NaN       NaN   
2000-01-04  0.010278 -0.038291  0.022766 -0.019340  0.000000 -0.020305   
2000-01-05 -0.004098  0.051619  0.037859  0.053082  0.017140  0.005115   
2000-01-06 -0.010320  0.045717 -0.009331  0.050405  0.001306 -0.008539   
2000-01-07  0.010320 -0.033436  0.000000 -0.002939 -0.002614 -0.022551   
...              ...       ...       ...       ...       ...       ...   
2022-12-23 -0.000553  0.014071  0.012770  0.026101  0.011057  0.000621   
2022-12-27  0.006782 -0.000635  0.008982  0.013798  0.007512  0.007728   
2022-12-28 -0.023464 -0.027055 -0.007229 -0.016563 -0.034579 -0.004552   
2022-12-29  0.032924  0.022277  0.007060  0.007537  0.045149  0.005245   
2022-12-30 -0.015306 -0.010270 -0.013600  0.010023 -0.013724 -0.014802   

                ZBRA  Year  
Date                        
2000-01-03       NaN  2000  
2000-01-04 -0.014533  2000  
2000-01-05  0.018963  2000  
2000-01-06 -0.055665  2000  
2000-01-07 -0.011160  2000  
...              ...   ...  
2022-12-23  0.002864  2022  
2022-12-27  0.011137  2022  
2022-12-28 -0.016713  2022  
2022-12-29  0.042396  2022  
2022-12-30 -0.004358  2022  

[5787 rows x 363 columns]

In [10]:
L_year_unique=list(np.unique(L_year))

df_annual_mean_return=pd.DataFrame(index=L_year_unique, columns=log_return_df.columns[:364])
df_annual_volatility=pd.DataFrame(index=L_year_unique, columns=log_return_df.columns[:364])
df_annual_sr=pd.DataFrame(index=L_year_unique, columns=log_return_df.columns[:364])

for stock in log_return_df.columns[:364]:
    L_mean=[]
    L_std=[]
    L_sr=[]
    for year in L_year_unique:
        data= log_return_df[log_return_df['Year']== year][stock] 
        m=data.mean()*252 # Mean
        std=data.std()*(252**0.5) # Volatility
        sr=m/std # Sharpe Ratio We make the hypothesis that the risk free rate is 0
        
        L_mean.append(m)
        L_std.append(std)
        L_sr.append(sr)
        
    df_annual_mean_return[stock]=L_mean
    df_annual_volatility[stock]=L_std
    df_annual_sr[stock]=L_sr

C:\Users\valou\AppData\Local\Temp\ipykernel_13096\2131628563.py:15: RuntimeWarning: divide by zero encountered in scalar divide
  sr=m/std # Sharpe Ratio We make the hypothesis that the risk free rate is 0


In [11]:
df_annual_mean_return

A      AAPL       ABT      ACGL      ADBE       ADI       ADM  \
2000 -0.274980 -1.330391  0.344510  0.266762  0.576717  0.127256  0.291818   
2001 -0.663047  0.393043  0.160215  0.549100 -0.636359 -0.144779  0.019149   
2002 -0.462108 -0.424130 -0.311069  0.191021 -0.223287 -0.620392 -0.129243   
2003  0.487390  0.399632  0.176237  0.245917  0.456352  0.649201  0.224037   
2004 -0.193325  1.103126  0.092239 -0.029534  0.474810 -0.206897  0.399683   
2005  0.323045  0.803171 -0.144830  0.346938  0.164180 -0.019095  0.115652   
2006  0.105733  0.166290  0.238892  0.211819  0.107083 -0.069109  0.271113   
2007  0.053024  0.851282  0.166542  0.039885  0.038560 -0.015318  0.388031   
2008 -0.851296 -0.838583 -0.025070 -0.003546 -0.693910 -0.477082 -0.458538   
2009  0.687051  0.903817  0.044774  0.020474  0.546717  0.540263  0.102650   
2010  0.287762  0.425711 -0.086308  0.207508 -0.178089  0.204556 -0.019503   
2011 -0.170659  0.227598  0.197015  0.237822 -0.085064 -0.024431 -0.029502   
2012  0.170061  0.284172  0.186256  0.168869  0.289627  0.194344 -0.018670   
2013  0.344522  0.077605  0.214762  0.304520  0.463213  0.220721  0.481794   
2014  0.007798  0.340909  0.183018 -0.009934  0.193999  0.114665  0.201533   
2015  0.034063 -0.030601  0.018545  0.165686  0.256315  0.023579 -0.324350   
2016  0.096788  0.117609 -0.131483  0.212796  0.091587  0.299797  0.249905   
2017  0.395758  0.396749  0.420565  0.050817  0.534032  0.226450 -0.099946   
2018  0.016582 -0.055630  0.256138 -0.124799  0.256458 -0.016593  0.051619   
2019  0.243257  0.636354  0.199508  0.473226  0.376920  0.345782  0.157880   
2020  0.333654  0.598147  0.246201 -0.172493  0.414686  0.237078  0.116794   
2021  0.303872  0.297495  0.266420  0.208903  0.125617  0.190323  0.317184   
2022 -0.057037 -0.307804 -0.232627  0.346647 -0.523857 -0.050571  0.337668   

           ADP      ADSK       AEE  ...       WMT       WRB       WST  \
2000  0.204163 -0.207606  0.432279  ... -0.225426  0.868137 -0.180412   
2001 -0.065737  0.338257 -0.030431  ...  0.087086  0.143144  0.111322   
2002 -0.395745 -0.254956  0.042796  ... -0.125162  0.110725 -0.053922   
2003  0.023240  0.548409  0.160730  ...  0.055755  0.288769  0.361467   
2004  0.126248  1.130162  0.140709  ...  0.005057  0.306420  0.411384   
2005  0.049066  0.124283  0.069189  ... -0.108671  0.420267  0.017282   
2006  0.087595 -0.059961  0.096702  ...  0.001331  0.088425  0.733765   
2007  0.022135  0.207722  0.058051  ...  0.047941 -0.140533 -0.222136   
2008 -0.094569 -0.925462 -0.422152  ...  0.181640  0.047776 -0.058503   
2009  0.119485  0.257065 -0.110513  ... -0.026172 -0.219339  0.054329   
2010  0.110208  0.407693  0.066402  ...  0.031688  0.115631  0.066890   
2011  0.183508 -0.230698  0.214215  ...  0.129666  0.237804 -0.065776   
2012  0.082580  0.154386 -0.025321  ...  0.158165  0.129202  0.385789   
2013  0.375578  0.353104  0.209614  ...  0.166901  0.148721  0.594644   
2014  0.185228  0.176941  0.283297  ...  0.112175  0.195696  0.090978   
2015  0.040189  0.014382 -0.023579  ... -0.309745  0.074801  0.131277   
2016  0.217448  0.194475  0.228563  ...  0.148689  0.221153  0.349774   
2017  0.153684  0.349527  0.148387  ...  0.383792  0.097638  0.157779   
2018  0.133792  0.205259  0.131490  ... -0.035061  0.058120 -0.000773   
2019  0.282979  0.355212  0.189093  ...  0.263617  0.363252  0.432632   
2020  0.056692  0.507416  0.041520  ...  0.208803 -0.031792  0.634322   
2021  0.354884 -0.082395  0.157800  ...  0.019517  0.242108  0.506081   
2022 -0.013004 -0.410245  0.025206  ... -0.004668  0.293261 -0.689872   

            WY       XEL       XOM      XRAY       YUM      ZBRA      Year  
2000 -0.289466  0.488715  0.126293  0.519619 -0.123310 -0.323595  504000.0  
2001  0.094453  0.006997 -0.080284  0.259854  0.405828  0.312925  504252.0  
2002 -0.065163 -0.849347 -0.092918  0.110715 -0.015568  0.031737  504504.0  
2003  0.292885  0.480636  0.187654  0.198963  0.350878  0.552410  504756.0  
20

In [12]:
df_annual_volatility

A      AAPL       ABT      ACGL      ADBE       ADI       ADM  \
2000  1.018473  1.013291  0.390895  0.315286  0.887646  0.936641  0.367685   
2001  0.681361  0.623043  0.282812  0.482446  0.822725  0.783454  0.341535   
2002  0.608804  0.489064  0.413288  0.302828  0.692421  0.661762  0.266271   
2003  0.473545  0.368728  0.236291  0.227445  0.408800  0.443563  0.220780   
2004  0.385757  0.396564  0.195655  0.188924  0.298697  0.328489  0.219449   
2005  0.266131  0.387665  0.186712  0.191458  0.328790  0.249658  0.343323   
2006  0.285786  0.381431  0.165499  0.164577  0.325087  0.304447  0.370860   
2007  0.265217  0.375966  0.193075  0.207634  0.243814  0.271433  0.278021   
2008  0.485070  0.587321  0.301897  0.392064  0.557204  0.481697  0.640327   
2009  0.410046  0.337117  0.258426  0.341765  0.461454  0.368908  0.407562   
2010  0.306790  0.266856  0.143468  0.159409  0.387890  0.293756  0.237203   
2011  0.489574  0.262498  0.176348  0.235846  0.337804  0.303221  0.334344   
2012  0.295244  0.293046  0.141326  0.146388  0.233939  0.197592  0.206057   
2013  0.219567  0.289052  0.194243  0.135622  0.246303  0.189076  0.217103   
2014  0.222704  0.216682  0.153932  0.120268  0.255751  0.210114  0.194471   
2015  0.225160  0.267491  0.218785  0.145107  0.235262  0.317752  0.238165   
2016  0.236124  0.233725  0.243751  0.162784  0.247617  0.236354  0.290918   
2017  0.165140  0.175283  0.139500  0.140958  0.214611  0.202277  0.188585   
2018  0.272158  0.287744  0.224907  0.197881  0.354572  0.274756  0.188017   
2019  0.247760  0.262970  0.199360  0.174727  0.238078  0.288778  0.202089   
2020  0.366090  0.466763  0.391134  0.577685  0.478132  0.503427  0.400742   
2021  0.207028  0.250920  0.213683  0.227350  0.296812  0.271764  0.208416   
2022  0.352293  0.356347  0.267216  0.277208  0.457323  0.367425  0.313235   

           ADP      ADSK       AEE  ...       WMT       WRB       WST  \
2000  0.379687  0.671304  0.253735  ...  0.516358  0.416492  0.330896   
2001  0.318565  0.573744  0.216840  ...  0.330858  0.364416  0.257596   
2002  0.439411  0.527905  0.242295  ...  0.320527  0.254858  0.277409   
2003  0.276854  0.372596  0.152086  ...  0.225604  0.220510  0.335306   
2004  0.204162  0.330814  0.127024  ...  0.165753  0.190637  0.272713   
2005  0.149478  0.394451  0.139973  ...  0.150864  0.210019  0.318023   
2006  0.151504  0.353162  0.121563  ...  0.169557  0.249265  0.309803   
2007  0.188190  0.284313  0.178652  ...  0.200896  0.254869  0.325906   
2008  0.392461  0.621189  0.410065  ...  0.333609  0.516861  0.427709   
2009  0.256732  0.482947  0.322156  ...  0.223264  0.369139  0.299357   
2010  0.172037  0.356675  0.187965  ...  0.139748  0.167770  0.240524   
2011  0.222967  0.456453  0.225010  ...  0.166110  0.246238  0.272157   
2012  0.141967  0.380698  0.141050  ...  0.163628  0.156944  0.217999   
2013  0.139560  0.271029  0.150119  ...  0.122626  0.132572  0.184256   
2014  0.144831  0.264934  0.182959  ...  0.132721  0.127245  0.225123   
2015  0.192863  0.283828  0.197845  ...  0.209652  0.181365  0.219079   
2016  0.170302  0.353428  0.183754  ...  0.190911  0.164234  0.226367   
2017  0.194350  0.305709  0.117948  ...  0.178819  0.135597  0.218163   
2018  0.228207  0.406801  0.184532  ...  0.240886  0.164307  0.267538   
2019  0.181778  0.293170  0.142502  ...  0.143115  0.164195  0.217764   
2020  0.448421  0.505029  0.432111  ...  0.311568  0.489363  0.391507   
2021  0.192425  0.349829  0.179888  ...  0.168123  0.215171  0.239915   
2022  0.283712  0.473961  0.219913  ...  0.270851  0.249852  0.426960   

            WY       XEL       XOM      XRAY       YUM      ZBRA  Year  
2000  0.450868  0.285123  0.314637  0.381945  0.498858  0.703657   0.0  
2001  0.323833  0.253148  0.245672  0.282925  0.296097  0.532601   0.0  
2002  0.392372  0.880484  0.321113  0.310668  0.428482  0.400347   0.0  
2003  0.237071  0.237574  0.175600  0.239007  0.255213  0.267874   0.0  
2004  0.212347  0.1432

In [13]:
df_annual_sr

A      AAPL       ABT      ACGL      ADBE       ADI       ADM  \
2000 -0.269993 -1.312940  0.881338  0.846093  0.649715  0.135865  0.793662   
2001 -0.973121  0.630843  0.566505  1.138159 -0.773476 -0.184796  0.056069   
2002 -0.759042 -0.867228 -0.752670  0.630791 -0.322473 -0.937485 -0.485380   
2003  1.029237  1.083812  0.745850  1.081218  1.116323  1.463606  1.014753   
2004 -0.501159  2.781708  0.471434 -0.156325  1.589602 -0.629844  1.821304   
2005  1.213858  2.071818 -0.775688  1.812082  0.499347 -0.076487  0.336861   
2006  0.369972  0.435963  1.443465  1.287048  0.329399 -0.226999  0.731039   
2007  0.199926  2.264256  0.862576  0.192094  0.158152 -0.056434  1.395689   
2008 -1.754998 -1.427811 -0.083043 -0.009044 -1.245344 -0.990419 -0.716100   
2009  1.675547  2.681016  0.173257  0.059906  1.184769  1.464491  0.251863   
2010  0.937977  1.595287 -0.601585  1.301733 -0.459122  0.696344 -0.082219   
2011 -0.348587  0.867046  1.117196  1.008381 -0.251815 -0.080570 -0.088240   
2012  0.576003  0.969719  1.317916  1.153571  1.238045  0.983563 -0.090607   
2013  1.569102  0.268481  1.105635  2.245364  1.880663  1.167361  2.219196   
2014  0.035015  1.573316  1.188956 -0.082595  0.758546  0.545726  1.036314   
2015  0.151282 -0.114399  0.084762  1.141822  1.089490  0.074204 -1.361872   
2016  0.409904  0.503195 -0.539416  1.307231  0.369874  1.268425  0.859022   
2017  2.396495  2.263483  3.014795  0.360510  2.488376  1.119500 -0.529977   
2018  0.060929 -0.193330  1.138864 -0.630678  0.723288 -0.060391  0.274542   
2019  0.981822  2.419873  1.000741  2.708373  1.583180  1.197398  0.781239   
2020  0.911398  1.281478  0.629455 -0.298593  0.867304  0.470927  0.291444   
2021  1.467784  1.185616  1.246801  0.918861  0.423221  0.700324  1.521880   
2022 -0.161902 -0.863775 -0.870557  1.250493 -1.145484 -0.137637  1.078003   

           ADP      ADSK       AEE  ...       WMT       WRB       WST  \
2000  0.537714 -0.309258  1.703666  ... -0.436568  2.084403 -0.545224   
2001 -0.206355  0.589561 -0.140339  ...  0.263213  0.392803  0.432159   
2002 -0.900626 -0.482958  0.176626  ... -0.390488  0.434457 -0.194377   
2003  0.083943  1.471860  1.056838  ...  0.247136  1.309550  1.078018   
2004  0.618369  3.416305  1.107738  ...  0.030509  1.607347  1.508485   
2005  0.328246  0.315078  0.494303  ... -0.720323  2.001090  0.054341   
2006  0.578165 -0.169783  0.795489  ...  0.007848  0.354743  2.368493   
2007  0.117622  0.730609  0.324941  ...  0.238637 -0.551392 -0.681597   
2008 -0.240965 -1.489822 -1.029477  ...  0.544470  0.092434 -0.136783   
2009  0.465406  0.532285 -0.343041  ... -0.117225 -0.594190  0.181484   
2010  0.640604  1.143038  0.353268  ...  0.226751  0.689225  0.278099   
2011  0.823030 -0.505415  0.952021  ...  0.780603  0.965749 -0.241685   
2012  0.581684  0.405535 -0.179515  ...  0.966610  0.823236  1.769685   
2013  2.691155  1.302829  1.396318  ...  1.361056  1.121810  3.227265   
2014  1.278927  0.667870  1.548412  ...  0.845197  1.537938  0.404124   
2015  0.208379  0.050670 -0.119178  ... -1.477425  0.412436  0.599225   
2016  1.276838  0.550253  1.243859  ...  0.778841  1.346575  1.545161   
2017  0.790761  1.143330  1.258078  ...  2.146263  0.720057  0.723220   
2018  0.586274  0.504569  0.712561  ... -0.145551  0.353728 -0.002889   
2019  1.556725  1.211623  1.326952  ...  1.841994  2.212316  1.986701   
2020  0.126425  1.004727  0.096087  ...  0.670168 -0.064967  1.620206   
2021  1.844274 -0.235530  0.877214  ...  0.116087  1.125190  2.109415   
2022 -0.045837 -0.865568  0.114616  ... -0.017234  1.173738 -1.615779   

            WY       XEL       XOM      XRAY       YUM      ZBRA  Year  
2000 -0.642020  1.714053  0.401391  1.360457 -0.247185 -0.459876   inf  
2001  0.291671  0.027639 -0.326795  0.918455  1.370593  0.587540   inf  
2002 -0.166076 -0.964637 -0.289362  0.356378 -0.036332  0.079275   inf  
2003  1.235435  2.023094  1.068643  0.832456  1.374845  2.062202   inf  
2004  0.351433  0.8068

# Question 4:

- Would it make sense to choose portfolio weights based only on the Sharpe ratios of the stocks in your dataset? Explain the reasons for your answer

To answer this question let calculate the skewness and the kurtosis

In [14]:
from scipy.stats import skew, kurtosis


df_skew=pd.DataFrame(index=L_year_unique, columns=log_return_df.columns[:364])
df_kurtosis=pd.DataFrame(index=L_year_unique, columns=log_return_df.columns[:364])


for stock in log_return_df.columns[:364]:
    L_skew=[]
    L_kurtosis=[]

    for year in L_year_unique:
        data= log_return_df[log_return_df['Year']== year][stock]
        skewness=skew(data)
        kurt=kurtosis(data)
       
        L_skew.append(skewness)
        L_kurtosis.append(kurt)

    df_skew[stock]=L_skew
    df_kurtosis[stock]=L_kurtosis

C:\Users\valou\AppData\Local\Temp\ipykernel_13096\2586983988.py:14: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.
  skewness=skew(data)
C:\Users\valou\AppData\Local\Temp\ipykernel_13096\2586983988.py:15: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.
  kurt=kurtosis(data)


In [15]:
df_skew

A      AAPL       ABT      ACGL      ADBE       ADI       ADM  \
2000       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
2001  0.230645 -0.062664 -0.639649  4.911366  0.088189  0.414626  0.014270   
2002  0.754271 -0.590528 -0.906509 -0.410577 -1.910866  0.444098 -0.743396   
2003 -3.561087  0.288410 -0.004780 -0.252865 -0.479714  0.267289  0.224166   
2004 -0.526210  1.125209  0.287035 -0.031003  0.529360 -0.042348  0.957776   
2005  2.272977 -0.124757 -0.925869  0.388781  0.686841  0.490038 -3.160653   
2006 -0.270169  0.704363  0.711093 -0.055851  1.070645 -2.972287  0.455602   
2007 -1.056564 -0.086215  0.599553  0.062388  0.008365  0.024833  0.843079   
2008  0.134649 -0.428121 -0.174109  0.103253 -0.043329 -0.054466  0.087744   
2009 -0.103255  0.210857 -0.379989 -0.132603  0.211886  0.816339 -1.049966   
2010 -0.151154  0.079636 -0.227585 -0.028189 -1.890373  0.079880 -0.325535   
2011 -0.306875 -0.118744 -0.159932  0.021375 -0.416737 -0.046649 -0.557199   
2012 -0.292103  0.443066 -0.200344 -0.708534  0.048813  0.552329  0.170808   
2013  0.561808 -1.500247  0.578059 -0.332369  2.578574 -0.154721 -0.319771   
2014 -0.822362 -0.363872 -0.092186 -0.677674  0.574901 -0.232879 -0.209031   
2015 -0.111375 -0.076136 -0.695202 -0.756361 -0.625151  0.776292 -0.608150   
2016 -0.810317 -0.321010 -1.879224  0.242110 -0.488176 -0.653579 -0.776230   
2017  0.548014  0.469543  0.211853 -0.737077  1.558371 -0.069729 -1.954091   
2018 -0.688610 -0.146497 -0.135486 -0.341032 -0.122383 -0.274484  0.084094   
2019 -1.840449 -1.128364 -0.752103 -0.205310 -0.151673 -0.192119 -0.677002   
2020 -0.302970 -0.308176  0.054711 -0.601065 -0.133598 -0.622629 -0.267776   
2021 -0.647179 -0.126143 -1.470461 -0.202862 -1.379983 -0.204127 -0.267437   
2022  0.314711  0.219090 -0.157695  0.443778 -1.014262  0.045033 -0.796700   

           ADP      ADSK       AEE  ...       WMT       WRB       WST  \
2000       NaN       NaN       NaN  ...       NaN       NaN       NaN   
2001 -0.202680  0.039922 -0.961699  ...  0.298803 -0.374540  0.649066   
2002 -3.447892 -0.773367  0.444226  ...  0.215955  0.652579  0.078297   
2003 -1.382260  1.065117  0.197221  ...  0.222725  0.044656 -0.494384   
2004  0.708374  0.660333 -0.384426  ... -0.011288 -0.439722  0.948229   
2005  1.355667 -2.439761 -0.287557  ...  0.329061  0.846387 -0.496628   
2006 -0.084201  0.684650 -0.215923  ...  0.213899  0.574684  0.667762   
2007  0.219481 -0.088584 -0.244412  ...  0.243759  0.478258 -0.237902   
2008  0.308762 -0.049669  0.671324  ...  0.367413  0.827044  0.338129   
2009 -0.035631 -0.271240 -3.475054  ... -0.380846 -0.168979 -0.082720   
2010  0.058641 -0.267204  0.125582  ... -0.363815  0.207892 -0.264174   
2011 -0.340130 -0.453274 -0.406313  ... -0.192961 -0.142434 -0.159829   
2012  0.030304 -2.000051  0.140604  ... -0.622479  0.582969  2.581035   
2013  0.189252  0.266497 -0.079729  ... -0.077665 -0.665872  0.248337   
2014 -0.179571  0.047579  0.150473  ...  0.357056  0.172804  0.565054   
2015 -0.133961  0.916588 -0.499877  ... -2.082806  0.053855 -0.204969   
2016 -0.309755 -0.379524 -0.677440  ...  1.301653 -0.033618 -0.645456   
2017 -0.222731 -1.682644 -0.363756  ...  2.830846 -0.794028  2.508331   
2018 -0.316588  1.133566 -0.410289  ... -0.683279 -0.232153  0.131139   
2019 -0.450252 -0.188465 -0.139307  ...  0.594684  0.278127  0.700972   
2020 -0.713585 -0.428777 -0.325250  ...  0.904438 -0.466029  0.146340   
2021 -0.366624 -2.216361 -0.517612  ... -1.131553 -0.273455 -0.721683   
2022 -0.362945  0.593027 -0.215264  ... -1.726614 -0.282072 -0.492099   

            WY       XEL       XOM      XRAY       YUM      ZBRA  Year  
2000       NaN       NaN       NaN       NaN       NaN       NaN   NaN  
2001 -0.367123 -0.444250 -0.005642 -0.631082  0.096763 -0.118143   NaN  
2002 -0.131931 -2.606260 -0.112113 -0.030852 -2.261327  0.833693   NaN  
2003 -0.066834  1.054287 -0.360792  0.420801  0.541089 -0.088958   NaN  
2004 -0.202932 -0.2474

In [16]:
df_kurtosis

A       AAPL        ABT       ACGL       ADBE        ADI  \
2000        NaN        NaN        NaN        NaN        NaN        NaN   
2001   0.682848   1.904782   2.895330  53.575741   2.707572   0.075429   
2002   2.690209   3.526898   9.415446   1.573106  16.444611  -0.040706   
2003  33.284592   2.137729   0.341153   2.684381   2.886436   1.061965   
2004   3.787389   3.737957   1.463303   1.880391   2.130718   0.035380   
2005  16.872024   1.437774   7.172376   1.639009   5.904455   1.592912   
2006   4.235464   2.021547   2.689098   1.757733   8.192329  27.526660   
2007  11.930598   2.003425   2.184462   0.920209   1.326705   8.932900   
2008   2.905112   3.261339   4.816253   4.408332   2.100953   2.815024   
2009   1.888634   0.492642   2.556943   1.536017   1.735251   5.302853   
2010   2.005290   1.982939   0.826906   1.002557  24.063219   0.499606   
2011   1.732533   1.166020   2.496045   4.504811   1.835471   0.526661   
2012   1.525231   2.842676   4.214262   2.241044   0.912062   0.800496   
2013   5.117251  10.609461   2.873460   1.373648  16.239148   1.109593   
2014   4.410640   9.202541   0.484070   2.586054   4.709419   4.929906   
2015   0.104262   1.313157   3.731621   3.587564   1.652345   4.786360   
2016   1.869343   4.330155   8.957659   2.141014   6.076370   3.451482   
2017   2.764201   4.564010   0.328825   3.975199  21.717125   2.277677   
2018   5.608544   1.845691   1.982999   2.680967   3.022764   1.297051   
2019  12.761424   8.080848   1.827975   0.632733   1.327801   1.211368   
2020   3.992567   3.823801   3.792255   4.714093   6.602130   7.633665   
2021   1.457535   0.334358  11.421209   0.504657   5.793491   0.478980   
2022   0.486406   0.840925   0.615789   2.828319   6.317846   0.240809   

            ADM        ADP       ADSK        AEE  ...        WMT       WRB  \
2000        NaN        NaN        NaN        NaN  ...        NaN       NaN   
2001   0.624567   2.358077   2.767364   3.585094  ...   1.219461  6.783563   
2002   3.086696  33.255964   6.119635   5.361626  ...   0.826282  2.194376   
2003   0.712727   7.953292   6.509160   2.492186  ...   1.024798  0.503521   
2004   4.831748   3.413701   5.759649   1.355217  ...   0.497545  3.492149   
2005  23.769016   5.946219  16.766943   0.947781  ...   1.463400  3.052005   
2006   1.861227   2.761419   2.415542  -0.025119  ...   0.590202  4.749441   
2007   4.452180   1.267485   1.262416   0.950386  ...   3.398945  5.306306   
2008   4.295636   3.508140   3.039004   6.944875  ...   4.298095  4.794283   
2009   5.258750   1.020677   2.587322  29.942740  ...   3.828222  0.972662   
2010   2.828742   1.365440   1.561063   0.607842  ...   1.375560  1.727297   
2011   1.601274   1.968365   1.980067   2.614045  ...   2.335506  1.793351   
2012   3.845070   1.249840  13.565814   1.507990  ...   4.221588  2.763609   
2013   1.603866   0.892388   2.333065   0.593336  ...   0.589466  3.174495   
2014   1.383487   1.832292   3.063338   1.016439  ...   4.963763  0.884975   
2015   2.444310   2.239020   4.606800   1.099458  ...  15.402397  2.455555   
2016   4.094405   1.906675   2.719455   1.352249  ...  12.640403  0.992071   
2017  15.168506  15.809633  34.599612   0.643951  ...  27.426605  5.288604   
2018   0.978586   2.159537   6.773218   3.500736  ...  13.843088  1.235668   
2019   4.964994   1.749976   1.358610   0.047208  ...   6.928035  4.664512   
2020   3.417123   7.704074   4.278326  10.724943  ...   8.588297  5.713274   
2021   0.910825   1.660721  13.704270   0.682790  ...   5.976667  1.023018   
2022   1.870274   3.533587   1.592333   1.020427  ...  12.158893  0.671811   

            WST        WY        XEL       XOM       XRAY        YUM  \
2000        NaN       NaN        NaN       NaN        NaN        NaN   
2001   6.008980  2.590887   2.222658  1.047666   2.584970   0.682677   
2002   3.378200  2.757686  22.358891  3.293179   3.143779  23.607804   
2003   1.761931  0.606909   4.913403  0.793695   1.893008   1.954555   
2004

# Sharpe Ratio Limitations:


- The Sharpe ratio has several limitations:

- Sharpe ratio is appropriate for efficient portfolios but not for individual assets because it ignores correlation between these assets.
- Sharpe ratio, because it depends on the mean and variance of returns, is not appropriate if returns have high skewness and kurtosis.
- Sharpe ratio looks at first two moments, but ignores the distribution of returns. In contrast, the Omega measure uses the entire distribution

- We are dealing with 363 individuals assets and we ignore the correlation between them.
- From our calculation we can see that Skewness and Kurtosis are quite high

So we conclude that choosing a portfolio weights based only on the Sharpe ratios of the stocks in our dataset will not be a good choice.